modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/source/poll/PollingSourceWrapper.java

```diff
@@ -253,25 +269,113 @@ public PollItemStatus accept(Consumer<PollItem<T, A>> consumer) {
 
       pollItem.validate();
 
-      PollItemStatus status;
+      PollItemStatus status = ACCEPTED;
+      boolean currentPollItemLimitApplied = false;
       if (!acquireItem(pollItem, callbackContext)) {
         status = ALREADY_IN_PROCESS;
-      } else if (!passesWatermark(pollItem)) {
-        status = FILTERED_BY_WATERMARK;
-      } else if (isRequestedToStop()) {
-        status = SOURCE_STOPPING;
       } else {
-        sourceCallback.handle(pollItem.getResult(), callbackContext);
-        status = ACCEPTED;
+        WatermarkStatus watermarkStatus = passesWatermark(pollItem);
+        if (watermarkStatus == REJECT) {
+          status = FILTERED_BY_WATERMARK;
+        } else if (isRequestedToStop()) {
+          status = SOURCE_STOPPING;
+        } else if (currentPollItems < maxItemsPerPoll) {
+          currentPollItems++;
+          sourceCallback.handle(pollItem.getResult(), callbackContext);
+          saveWatermarkValue(watermarkStatus, pollItem);
+        } else {
+          currentPollItemLimitApplied = true;
+          processLimitApplied(watermarkStatus, pollItem);
+        }
       }
 
-      if (status != ACCEPTED) {
+      if (status != ACCEPTED || currentPollItemLimitApplied) {
         rejectItem(pollItem.getResult(), callbackContext);
       }
 
       return status;
     }
 
+    private void processLimitApplied(WatermarkStatus watermarkStatus, DefaultPollItem pollItem) {
+      Serializable itemWatermark = pollItem.getWatermark().orElse(null);
+      if (itemWatermark == null || watermarkStatus != PASSED) {
+        return;
+      }
+      if (minimumRejectedByLimitPassingWatermark == null ||
+          compareWatermarks(itemWatermark, minimumRejectedByLimitPassingWatermark, watermarkComparator) < 0) {
+        LOGGER.debug("An item that passed all previous validations is being rejected by the poll limit and its watermark" +
+            "value will be stored so that is processed on future polls if sent for processing.");
+        minimumRejectedByLimitPassingWatermark = itemWatermark;
+      }
+    }
+
+    private void saveWatermarkValue(WatermarkStatus watermarkStatus, DefaultPollItem pollItem) {
```

Comment:
isn't this change actually tackling the watermark issues rather than implementing the limit thing? If the answer is yes, then split into different PRs

Reply:
For the limit feature, I need to test watermarking without impacting the object stores, after that, if the item is not rejected due to the poll limit, the pertinent values will be stored. This is why, the passesWatermark method returns a WatermarkStatus, and later using that status we know what to store in the object stores.